Modules/user.py

```diff
@@ -0,0 +1,139 @@
+"""
+user.py - User objects main program
+
+Copyright (c) 2018 The Fuel Rats Mischief,
+All rights reserved.
+
+Licensed under the BSD 3-Clause License.
+
+See LICENSE.md
+
+This module is built on top of the Pydle system.
+
+"""
+import logging
+from typing import Union
+
+from pydle import BasicClient
+
+import config
+from Modules import permissions
+from Modules.permissions import Permission
+
+LOG = logging.getLogger(f"{config.Logging.base_logger}.Permissions")
+
+
+class User(object):
+    """Represents an IRC user"""
+
+    def __init__(self,
+                 realname: str,
+                 hostname: str,
+                 username: str,
+                 away: bool,
+                 away_message: Union[str, None],
+                 account: str,
+                 identified: bool = False):
+        """
+        Create a new IRC User representation
+
+        Args:
+            realname (str): real name
+            hostname (str): hostmask
+            username (str): username
+            away (bool): user's away status
+            away_message(str): user's away message
+            account (?): ?
+            identified (bool): user identification status against IRC services
+
+        """
+        self._realname: str = realname
+        self._hostname: str = hostname
+        self._username: str = username
+        self._identified: bool = identified
+        self._away: bool = away
+        self._account: str = account
+        self._permission_level: Permission = permissions.NONE
+        self._away_message: Union[str, None] = away_message
+        # sets the permission based on the hostmask
+        # which requires stripping the username and the leading period
+        try:
+            # FIXME: access to protected member
+            # loop over every known vhost
+            for key, value in permissions._by_vhost.items():
+                # if a vhost matches, and the user is identified
+                if hostname.endswith(key) and identified is True:
+                    self._permission_level = value
+        except KeyError:
```

Comment:
Hey, I see you added the permission level as a property of `User`. I have two concerns with this:

1. As you point out yourself, this accesses the private `_by_vhost` directly. Simply creating a public, top-level function `has_level(vhost: str, level:Permission)` in `permissions.py` will easily solve this.

2. Why not just leave it as it was? Letting the `require_permission` decorator: a) worked fine, so why try (and fail) to fix it? b) avoids adding another public member to the permissions module

I suppose there could at some point be some sort of reason for a command function to inquire the users permission after it has been called, however I can't come up with a circumstance for this and, as you were so intent on in the `Rescue.outcome` discussion, where our opinions were reversed, why build what we don't need (and will likely never need)?

This also brings up another, deeper problem I see in development style (and this is not at all the only instance of this): Adding TODOs and FIXMEs to code which you know to be broken does not make it any better. In fact, it gives an impression, to me at least, that you didn't care to fix the issue itself. You moved the JIRA story on this to 'In review'. That leads me to believe that you consider this done and up for merging. So why is it still broken, especially when the solution is not that hard? I ask you to strive towards cleaner and better polish here. If not before pushing, as I believe would be ideal, then before peer review.

I hope you don't see me saying that as rude, because I don't want it to be and I have tried to keep it polite, however I also don't believe in keeping quiet about these things as this does bug me.

Reply:
It was a "this might be useful" kinds of things, to which i forgot was already implemented elsewhere. You are absolutely right for calling me on this. the `require_permission` decorators job is to handle user permissions, i should not be reinventing the wheel.

It is that i do not care to fix a problem itself, more that i know its broken but i cannot fix it without things that do not yet exist. I have reviewed all my TODOs and FIXMEs and all bot two of them require at least the API interface to exist. (after removing items that are no longer valid.) 
It is NOT that i do not care to fix issues, it is that i know there are issues that i cannot resolve without making fatal assumptions dependent on features that do not exist within the codebase (yet). Sure i can peak over to your branch and write the associated call, but i would rather wait until your branch is PRed in so you can make any changes you need prior to pushing without needing to inform you made a breaking change.

AS for why i marked the feature as "in review" as the first part of the large story (the Context object) is what i consider to be ready for review. Granted i should have created a task rather than moving the whole story, but i was under the impression the changes i made to the title and description of this PR were sufficient to communicate that fact. Evidently i am mistaken and i took the incorrect path to expressing my intentions.

corrections incoming.